python/lsst/daf/butler/dimensions/_record_table.py

```diff
@@ -0,0 +1,232 @@
+
+    Parameters
+    ----------
+    element : `DimensionElement` or `str`, optional
+        The dimension element that defines the records held by this table. If
+        not a `DimensionElement` instance, ``universe`` must be provided.  If
+        not provided, ``table`` must have an "element" entry in its metadata
+        (as is the case for tables returned by the `to_arrow` method).
+    records : `~collections.abc.Iterable` [ `DimensionRecord` ], optional
+        Dimension records to add to the table.
+    universe : `DimensionUniverse`, optional
+        Object that defines all dimensions.  Ignored if ``element`` is a
+        `DimensionElement` instance.
+    table : `pyarrow.Table`
+        Arrow table to copy columns from.  Must have schema returned by
+        `make_arrow_schema` for this element.  This argument is primarily
+        intended to serve as the way to reconstruct a `DimensionRecordTable`
+        that has been serialized to an Arrow-supported file or IPC format.
+
+    Notes
+    -----
+    `DimensionRecordTable` should generally have a smaller memory footprint
+    than `DimensionRecordSet` if its rows are unique, and it provides fast
+    column-oriented access and Arrow interoperability that `DimensionRecordSet`
+    lacks entirely.  In other respects `DimensionRecordSet` is more
+    featureful and simpler to use efficiently.
+    """
+
+    def __init__(
+        self,
+        element: DimensionElement | str | None = None,
+        records: Iterable[DimensionRecord] = (),
+        universe: DimensionUniverse | None = None,
+        table: pa.Table | None = None,
+    ):
+        if element is None:
+            if table is not None and b"element" in table.schema.metadata:
+                element = table.schema.metadata[b"element"].decode()
+            else:
+                raise TypeError("If 'element' is not provided it must be present in 'table.schema.metadata'.")
+        if isinstance(element, str):
+            if universe is None:
+                raise TypeError("'universe' must be provided if 'element' is not a DimensionElement.")
+            element = universe[element]
+        else:
+            universe = element.universe
+        self._element = element
+        self._converters = element.schema.to_arrow()
+        arrow_schema = pa.schema(
+            [converter.field for converter in self._converters],
+            {
+                b"element": element.name.encode(),
+                # Since the Arrow table might be saved to a file on its own, we
+                # include the dimension universe's identifiers in its metadata.
+                b"namespace": element.universe.namespace.encode(),
+                b"version": str(element.universe.version).encode(),
+            },
+        )
+        self._required_value_fields = [pc.field(name) for name in self._element.schema.required.names]
+        batches = [self._make_batch(records, arrow_schema)]
```

Comment:
It might be worth breaking the single batch up into multiple smaller batches -- especially if `records` is sometimes incrementally generated.

This would reduce peak memory usage by not having a duplicate copy of the full set of data in Python lists, as well as reducing the maximum size of the temporary string columns that will eventually be converted to dictionaries.

Reply:
Good point.  I don't have any intuition for what a default chunk size is, so I'll make it an optional `__init__` parameter that defaults to `None` and hence defaults to the current behavior.